pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/CoLocatedOfflineSegmentAssignment.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.pinot.common.metadata.ZKMetadataProvider;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.common.tier.Tier;
+import org.apache.pinot.segment.spi.partition.metadata.ColumnPartitionMetadata;
+import org.apache.pinot.spi.config.table.ReplicaGroupStrategyConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class CoLocatedOfflineSegmentAssignment implements SegmentAssignment {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CoLocatedOfflineSegmentAssignment.class);
+
+  private HelixManager _helixManager;
+  private String _offlineTableName;
+  private int _replication;
+  private String _partitionColumn;
+
+  @Override
+  public void init(HelixManager helixManager, TableConfig tableConfig) {
+    _helixManager = helixManager;
+    _offlineTableName = tableConfig.getTableName();
+    _replication = tableConfig.getValidationConfig().getReplicationNumber();
+    ReplicaGroupStrategyConfig replicaGroupStrategyConfig =
+        tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    _partitionColumn = replicaGroupStrategyConfig != null ? replicaGroupStrategyConfig.getPartitionColumn() : null;
+
+    if (_partitionColumn == null) {
+      LOGGER.error("Co-located segment assignment works only when a segment partition column is specified");
+    } else {
+      LOGGER.info("Initialized OfflineSegmentAssignment with replication: {} and partition column: {} for table: {}",
+          _replication, _partitionColumn, _offlineTableName);
+    }
+  }
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+    SegmentZKMetadata segmentZKMetadata = ZKMetadataProvider
+        .getSegmentZKMetadata(_helixManager.getHelixPropertyStore(), _offlineTableName, segmentName);
+    Preconditions
+        .checkState(segmentZKMetadata != null, "Failed to find segment ZK metadata for segment: %s of table: %s",
+            segmentName, _offlineTableName);
+    int segmentPartitionId = getPartitionId(segmentZKMetadata);
+    return SegmentAssignmentUtils.assignSegmentForCoLocatedTable(
+        instancePartitionsMap.get(InstancePartitionsType.OFFLINE), segmentPartitionId);
+  }
+
+  @Override
+  public Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, String>> currentAssignment,
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap, @Nullable List<Tier> sortedTiers,
+      @Nullable Map<String, InstancePartitions> tierInstancePartitionsMap, Configuration config) {
+    InstancePartitions instancePartitions = instancePartitionsMap.get(InstancePartitionsType.OFFLINE);
```

Comment:
how do we handle offline/online tyeps differently?

Reply:
At the moment both Offline and Realtime segment assignment rebalance are kinda the same.

For realtime tables the rebalance operation doesn't work right now.. I think it's because consuming segments will need to be handled separately and can't be treated the same as completed realtime segments. I am hoping to fix this by EOW.